StRoot/StGenericVertexMaker/StGenericVertexFinder.cxx

```diff
@@ -294,21 +294,28 @@ double StGenericVertexFinder::CalcChi2Beamline(const StThreeVectorD& point)
    double ky_dy_zv_2 = ky_dy_zv*ky_dy_zv;
    double kx_dx_zv_2 = kx_dx_zv*kx_dx_zv;
 
-   double denom_sqrt = kx2_ky2_1 * sqrt( ( (kx*dy - ky*dx)*(kx*dy - ky*dx) + (ky*zv - dy)*(ky*zv - dy) + (kx*zv - dx)*(kx*zv - dx) ) /kx2_ky2_1);
-
-   // The denominator is zero when the point is exactly on the beamline
-   // We just return a zero for the chi2 in this case. This makes sense for all
-   // non-zero errors and if they are zero they are unphysical anyway.
-   if (denom_sqrt == 0) return 0;
-
    // The distance between the line and the point
    StThreeVectorD dist_vec (
       (  (ky2 + 1)*dx -     kx*ky*dy -          kx*zv)/kx2_ky2_1,
       (    - kx*ky*dx + (kx2 + 1)*dy -          ky*zv)/kx2_ky2_1,
       (       - kx*dx -        ky*dy + (kx2 + ky2)*zv)/kx2_ky2_1
    );
 
+   double dist_mag = dist_vec.mag();
 
+   // When the vertex (point) gets closer to the beamline there may not be
+   // enough precision to do the intermediate calculations for the
+   // transformation of the errors, i.e. the Jacobian below. The distance is
+   // measured in cm so, anything within a nanometer will be considered to be on
+   // the beam line and we just return a zero for the chi2 in such cases. This
+   // makes sense for all non-zero errors (that is what we actually expect and
+   // assume) and if they are zero they are unphysical anyway.
+   if (dist_mag < 1e-7) return 0;
+
+   // A note for future optimizers: The expression sqrt(...) below is actually
```

Comment:
Is this a suggestion to optimize? Or to not touch the code? Manual optimizations better be put towards readability and maintainability, unless there is an indication that the compiler can't reasonably handle speed optimizations.

Reply:
This is similar to a case where you calculate the same quantity using two different math expressions, e.g. A = (x-y)(x+y) and B = x*x - y*y. When both results A and B are used the compiler won't optimize, although mathematically equivalent these can lead to different numerical results. The latter is the reason why I wouldn't touch the code now without a proper test demonstrating that the final result does not change much.